### PR TITLE
--pretty-er output by default

### LINKED ISSUE #10488: `--pretty` style errors should be the default on supported TTYs
Title. It would be nice to add a `--errorStyle [pretty|terse|auto]` command and deprecate `--pretty` (which would then be equivalent to `--errorStyle pretty`) as well.

cc @DanielRosenwasser

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This pull request always makes TypeScript's output `--pretty` when the compiler can detect whether its "output device" is appropriate for more colorful output.

If TypeScript's output is a psuedo-TTY, then it will enable today's prettified output and emit escape characters to the console; however, when piping to another file or program, it will automatically turned off.

Programs that want to specify the behavior can specifically set `--pretty`, or the new `--diagnosticStyle` flag.

This PR also affords us the room to give more variation in out emit modes. For example, with `--diagnosticStyle`, we could have a more colorful simple/terse mode as requested in #10745.

Fixes #10488.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
I know the last time I did this (i.e. when I originally wrote `--pretty`), there were issues with either Cygwin's bash terminal or MinGW. I forgot which, but I'll have to peek at that again.

Tests are Monday-Daniel's problem. 😄

File: src/compiler/commandLineParser.ts [hunk 1]
```diff
@@ -56,6 +56,14 @@ namespace ts {
             category: Diagnostics.Command_line_Options,
             description: Diagnostics.Stylize_errors_and_messages_using_color_and_context_experimental
         },
+        {
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: I would rather we do not add a new flag if we can. i think `auto` is just `pretty === undefined`. 

if we really need a new value i would make pretty take `boolean | "auto"`

File: src/compiler/tsc.ts [hunk 2]
```diff
@@ -19,11 +19,18 @@ namespace ts {
...
+        if (shouldBePretty(options)) {
             reportDiagnostic = createDiagnosticReporter(sys, /*pretty*/ true);
         }
     }
 
+    function shouldBePretty(options: CompilerOptions) {
+        if ((typeof options.pretty === "undefined" && typeof options.diagnosticStyle === "undefined") || options.diagnosticStyle === DiagnosticStyle.Auto) {
+            return !!sys.writeOutputIsTty && sys.writeOutputIsTty();
+        }
+        return options.diagnosticStyle === DiagnosticStyle.Pretty || options.pretty;
```
Comment (mhegazy) [inline src/compiler/tsc.ts]: !! options.pretty

File: src/compiler/commandLineParser.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/commandLineParser.ts]: `boolean | "auto" | "simple" | "styled"`? (Where boolean toggles between `auto` and `styled`).

File: src/compiler/sys.ts [hunk 3]
```diff
@@ -428,6 +428,7 @@ namespace ts {
         newLine: string;
         useCaseSensitiveFileNames: boolean;
         write(s: string): void;
+        writeOutputIsTty?(): boolean;
```
Comment (DanielRosenwasser) [inline src/compiler/sys.ts]: Is this a reasonable name @weswigham?

Comment (weswigham) [reply src/compiler/sys.ts]: `TTY`, not `Tty`.

File: src/compiler/tsc.ts [hunk 4]
```diff
@@ -19,11 +19,18 @@ namespace ts {
...
     let reportDiagnostic = createDiagnosticReporter(sys);
     function updateReportDiagnostic(options: CompilerOptions) {
-        if (options.pretty) {
+        if (shouldBePretty(options)) {
             reportDiagnostic = createDiagnosticReporter(sys, /*pretty*/ true);
         }
     }
 
+    function shouldBePretty(options: CompilerOptions) {
+        if ((typeof options.pretty === "undefined")) {
```
Comment (mhegazy) [inline src/compiler/tsc.ts]: extra parens.

**Comment by DanielRosenwasser** [comment]
Tried it out. tsc.exe (the Chakra host-wrapped `tsc`) reverts to the old behavior. Like before, `--pretty` on `tsc.exe` produces garbled text when explicitly specified, but that's always been the behavior.

**Comment by mhegazy** [comment]
@DanielRosenwasser please update https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Compiler%20Options.md as well.

**Comment by tangooncash** [comment]
Do you get paid to decode or put data together at github

**Comment by DanielRosenwasser** [comment]
Thanks for the pointers, I might look into it @kpreisser!

**Pull request merged**